micro-rdk-nmea/src/messages/pgns.rs

```diff
@@ -106,3 +115,96 @@ pub struct GnssSatsInView {
     #[length_field = "sats_in_view"]
     satellites: Vec<Satellite>,
 }
+
+macro_rules! define_pgns {
+    ( $(($pgndef:ident, $enum:ident, $pgn:expr)),* ) => {
+        #[derive(Clone, Debug)]
+        pub enum MessageData {
+            $($enum($pgndef)),*,
+            Unsupported(UnparsedMessageData)
+        }
+
+        impl MessageData {
+            pub fn pgn(&self) -> u32 {
+                match self {
+                    $(Self::$enum(_) => $pgn),*,
+                    Self::Unsupported(unparsed) => unparsed.pgn()
+                }
+            }
+
+            pub fn from_bytes(pgn: u32, bytes: Vec<u8>) -> Result<Self, crate::parse_helpers::errors::NmeaParseError> {
+                Ok(match pgn {
+                    $($pgn => {
+                        let cursor = DataCursor::new(bytes);
+                        Self::$enum($pgndef::from_cursor(cursor)?)
+                    }),*,
+                    x => Self::Unsupported(UnparsedMessageData::from_bytes(bytes, x)?)
+                })
+            }
+
+            pub fn to_readings(self) -> Result<GenericReadingsResult, crate::parse_helpers::errors::NmeaParseError> {
+                match self {
+                    $(Self::$enum(msg) => msg.to_readings()),*,
+                    Self::Unsupported(msg) => msg.to_readings()
+                }
+            }
+        }
+    };
+}
+
+define_pgns!(
+    (WaterDepth, Pgn128267, 128267),
+    (TemperatureExtendedRange, Pgn130316, 130316),
+    (GnssSatsInView, Pgn129540, 129540)
+);
+
+pub struct NmeaMessage {
+    metadata: NmeaMessageMetadata,
+    data: MessageData,
+}
+
+impl NmeaMessage {
+    pub fn new(mut bytes: Vec<u8>) -> Result<Self, NmeaParseError> {
```

Comment:
I'm curious why this takes the `Vec` by value rather than just viewing the data, since ultimately, everything gets parsed out, shouldn't this be able to take something like a byte slice, all the way down?

Should this implement `TryInto` at the top level?

Reply:
I do see what you are saying about the virality of lifetime when dealing with byte slices, but, on the other hand, `Vec` is saying that this method needs a uniquely owned buffer of data in order to convert it into an `NmeaMessage`. That seems like it shouldn't be needed. If I have a bunch of bytes that I read into a buffer off the wire or bus that is, say, already in a Vec, I'd expect the parser to just be able to consume that data by read-only reference, rather than needing to copy it into a Vec just to call this method. On the other hand, I don't want to delay this work further. Would you be OK with filing a follow-up ticket?

Reply:
I previously had `DataCursor` work this way, but then @npmenard said it should own the data. If you can convince him this is better when he gets back, I'll make the ticket.

My opinion on this is not strong but I don't really see a use-case outside of debugging to want a read-only reference to the data meant to be converted. I can see a use case for maybe supporting any type that implements Read. I could make a ticket for that if you agree